content/blog/2024/09/usr-merge/index.md

```diff
@@ -0,0 +1,90 @@
+
+### Sounds like Stateless
+
+If you've used Solus for any length of time, you may have heard the team talking about "stateless" packages. What does that mean? A stateless system means that, in theory, you can remove everything in `/etc` and be left with a still-working system using vendor-provided (e.g., Solus) defaults. We do this by putting default configuration files in `/usr/share/defaults`, which can be copied to `/etc` or `$HOME/.config` and modified by the user, overwriting the system defaults. While not directly related to Statelessness, the idea behind it and Usr Merge are quite similar.
+
+This concept is also called "Hermetic-Usr" by systemd and openSUSE.
+
+Serpent OS and `moss` takes the idea of Stateless a step further. The package manager *cannot* install files outside of `/usr`, enforcing a stateless system. Since Solus will also be moving to `moss` and associated tooling, we will also not be able to ship files outside of `/usr`, including files in directories like `/bin` and `/lib`. That's where Usr Merge comes in; it is a prerequisite to switching to the Serpent tooling on Solus.
+
+## Our previous attempt
+
+Back in May 2024, we thought we had a safe way to merge `/bin` and `/sbin` into `/usr`. It was discovered that we could simply turn these directories into symlinks in our `baselayout` package, and everything in them would end up in `/usr/bin` and `/usr/sbin` when they were updated. As it turned out, it was not that simple at all.
+
+There were two issues with this approach.
+
+1) Package updates with `eopkg` are non-deterministic. That means that there is no guarantee about the order that packages will be installed in. In this case, `baselayout` needed to be updated before any other package to create the symlinks, but there was no way to actually enforce it to be the first in the list. That created the situation where other packages were updated first, and thus their executable files didn't end up in `/usr/bin` or `/usr/sbin`, causing people's systems to be completely broken.
+2) The Usr Merge work depended on a previous update already having been installed on user's systems. When fetching updates, `eopkg` goes straight from where you are to the latest package version found in the repository, skipping anything in between. Since the team's systems are usually always up-to-date on the Unstable repository, we didn't catch that if you weren't already up-to-date, you would be missing the necessary work done in a previous update when you tried to install the Usr Merge update.
+
+The second issue is compounded by the first issue because even if the second is solved, the first makes it impossible to reliably install an update addressing it. We were stuck in the position where some users were successfully Usr Merged, other users weren't, and there was no way to unify those two groups, either by successfully merging everyone, or rolling back the merge.
+
+We started calling it the Chicken and the Egg problem.
```

Comment:
This reads a bit weird to me, as if the Solus team claims to have invented the "Chicken and/or the egg" phrase.

Also it's called "Chicken and the Egg" here, then "Chicken and the **e**gg" in the next header, and "Chicken and Egg" in the next section. Best to unify that.

Reply:
Re-worded slightly